### PR TITLE
support initial scheduler configuration option via an environment variable

### LINKED ISSUE #37: pass an initial scheduler configuration from CLI
Users can change the scheduler configuration through Web UI. 
But it would be nice if we can pass an initial scheduler configuration when we start the simulator.

/kind feature

### PR DESCRIPTION
**Pull request by momom-i**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add initial scheduler configuration from file path taken from environment variable.
#### Which issue(s) this PR fixes:

Fixes #37

#### Special notes for your reviewer:
I've confirmed that it also worked when I passed configuration file path.
I'd like to learn more so correcting my every little mistake is welcome!

/label tide/merge-method-squash

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -35,10 +38,16 @@ func NewConfig() (*Config, error) {
 		return nil, xerrors.Errorf("get frontend URL: %w", err)
 	}
 
+	initialschedulerCfg, err := getSchedulerCfg()
+	if err != nil {
+		return nil, xerrors.Errorf("get SchedulerConfigPath: %w", err)
```
Comment (sanposhiho) [inline config/config.go:43]:
```suggestion
		return nil, xerrors.Errorf("get SchedulerCfg: %w", err)
```

File: config/config.go [hunk 2]
```diff
@@ -73,3 +82,43 @@ func getFrontendURL() (string, error) {
 
 	return e, nil
 }
+
+// getSchedulerCfg read from kube scheduler config file path environment variable
+// named KUBE_SCHEDULER_CONFIG_PATH and convert it into *v1beta2config.KubeSchedulerConfiguration.
```
Comment (sanposhiho) [inline config/config.go:87]: nits:

```suggestion
// getSchedulerCfg reads KUBE_SCHEDULER_CONFIG_PATH which means initial kube-scheduler configuration 
// and converts it into *v1beta2config.KubeSchedulerConfiguration.
// KUBE_SCHEDULER_CONFIG_PATH is not required. 
// If KUBE_SCHEDULER_CONFIG_PATH is not set, the default configuration of kube-scheduler will be used.
```

File: config/config.go [hunk 3]
```diff
@@ -73,3 +82,43 @@ func getFrontendURL() (string, error) {
...
+func getSchedulerCfg() (*v1beta2config.KubeSchedulerConfiguration, error) {
+	e := os.Getenv("KUBE_SCHEDULER_CONFIG_PATH")
+	if e == "" {
+		return nil, nil
```
Comment (sanposhiho) [inline config/config.go:91]: I think it's better to return the default configuration here if KUBE_SCHEDULER_CONFIG_PATH is not set.

If we do so, we can remove this kinda logic.
- https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/80/files#diff-8f41374cc7da8c5d92c41034deebe762cc34593a7496a0000646715ab75cbd16R56-R62
- https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/80/files#diff-4741d40c952e5146eb01d0005e03250ada8a9d57c375f9f7c7026ab53cb1f7e1R50-R56

**Review by sanposhiho — COMMENTED** [review]
Nice work. 
Most of the changes look good. I left a few small comments.

File: config/config.go [hunk 4]
```diff
@@ -73,3 +82,43 @@ func getFrontendURL() (string, error) {
...
+func decodeSchedulerCfg(buf []byte) (*v1beta2config.KubeSchedulerConfiguration, error) {
+	decoder := scheme.Codecs.UniversalDeserializer()
+	obj, _, err := decoder.Decode(buf, nil, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("load an k8s object from buffer: %w", err)
+	}
+
+	sc, ok := obj.(*v1beta2config.KubeSchedulerConfiguration)
+	if !ok {
+		return nil, xerrors.Errorf("convert to KubeSchedulerConfiguration, but got unexpected type: %T", obj)
```
Comment (sanposhiho) [inline config/config.go:116]:
```suggestion
		return nil, xerrors.Errorf("convert to *v1beta2config.KubeSchedulerConfiguration, but got unexpected type: %T", obj)
```

**Comment by sanposhiho** [comment]
And could you sign cla?
https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/80#issuecomment-1012050774

File: config/config.go (hunk 3 above)
Comment (momom-i) [reply config/config.go:93]: ah, I missed it! Thank you, I'll fix it.

**Comment by momom-i** [comment]
a

**Comment by momom-i** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](mailto:login-issues@jira.linuxfoundation.org)

new comment

**Comment by momom-i** [comment]
I signed it

**Comment by momom-i** [comment]
I signed it.

File: config/config.go [hunk 5]
```diff
@@ -73,3 +84,50 @@ func getFrontendURL() (string, error) {
...
+
+// getSchedulerCfg reads KUBE_SCHEDULER_CONFIG_PATH which means initial kube-scheduler configuration
+// and converts it into *v1beta2config.KubeSchedulerConfiguration.
+// KUBE_SCHEDULER_CONFIG_PATH is not required.
+// If KUBE_SCHEDULER_CONFIG_PATH is not set, the default configuration of kube-scheduler will be used.
+func getSchedulerCfg() (*v1beta2config.KubeSchedulerConfiguration, error) {
+	dsc, err := defaultconfig.DefaultSchedulerConfig()
+	if err != nil {
+		return nil, xerrors.Errorf("create default scheduler config: %w", err)
+	}
```
Comment (sanposhiho) [inline config/config.go:96]: nit: Move these lines after L99(`if e == ""`)

Comment (momom-i) [reply config/config.go]: fixed: 3a0a5aa

**Review by sanposhiho — APPROVED** [review]
Thanks for the great work!! 
As a next step, could you add an explanation in README about this feature in another PR?

/lgtm
/approve

**Comment by sanposhiho** [comment]
/retitle support initial scheduler configuration option via an environment variable

**Pull request merged**